File: app/src/main/kotlin/com/test/business/account/Account.kt
package com.test.business.account

import com.test.CustomerId
import java.math.BigDecimal

typealias AccountId = String
typealias Iban = String

sealed interface Account : AccountInfo {
  val accountId: AccountId
  val customerId: CustomerId
  val isDefault: Boolean
  val balanceCount: Int
  val isPaymentAllowed: Boolean

  data class CurrentAccount(
    override val accountId: AccountId,
    override val customerId: CustomerId,
    override val iban: Iban,
    override val alias: String?,
    override val isDefault: Boolean,
    override val balanceCount: Int,
    override val isPaymentAllowed: Boolean
  ) : Account

  data class SavingsAccount(
    override val accountId: AccountId,
    override val customerId: CustomerId,
    override val iban: Iban,
    override val alias: String?,
    override val isDefault: Boolean,
    override val balanceCount: Int,
    override val isPaymentAllowed: Boolean,
    val goal: SavingsGoal?,
    val isRoundupEnabled: Boolean
  ) : Account {
    override val name: String
      get() = goal?.description ?: alias ?: iban
  }
}

data class SavingsGoal(
  val description: String?,
  val amount: BigDecimal?
)

File: ios/Sources/ios/Business/Account/Account.swift
typealias AccountId = String
typealias Iban = String
typealias MonetaryNumber = String

protocol AccountInfo {
    var iban: Iban { get }
    var alias: String? { get }
    var name: String { get }
}

protocol BaseAccount: AccountInfo {
    var accountId: AccountId { get }
    var customerId: CustomerId { get }
    var isDefault: Bool { get }
    var balanceCount: Int { get }
    var isPaymentAllowed: Bool { get }
}

enum AccountType {
    case current, credit, savings
}

@dynamicMemberLookup
enum Account: Equatable {
    case current(CurrentAccount)
    case savings(SavingsAccount)
    
    subscript<T>(dynamicMember keyPath: KeyPath<BaseAccount, T>) -> T {
        switch self {
        case let .current(account):
            return account[keyPath: keyPath]
        case let .savings(account):
            return account[keyPath: keyPath]
        }
    }
    
    var isSavingsAccount: Bool {
        if case .savings = self {
            return true
        }
        return false
    }
}

struct CurrentAccount: BaseAccount, Equatable {
    let accountId: AccountId
    let customerId: CustomerId
    let isDefault: Bool
    let balanceCount: Int
    let isPaymentAllowed: Bool
    let iban: Iban
    let alias: String?
}

struct SavingsAccount: BaseAccount, Equatable {
    let accountId: AccountId
    let customerId: CustomerId
    let isDefault: Bool
    let balanceCount: Int
    let isPaymentAllowed: Bool
    let iban: Iban
    let alias: String?
    let goal: SavingsGoal?
    let isRoundUpEnabled: Bool
    
    var name: String { goal?.description ?? alias ?? iban }
}

struct SavingsGoal: Equatable {
    let amount: MonetaryNumber?
    let description: String?
}

extension AccountInfo {
    var name: String { alias ?? iban }
}
